src/os_auth/main-server.c

```diff
@@ -739,6 +741,59 @@ void* run_dispatcher(__attribute__((unused)) void *arg) {
                 continue;
             }
 
+            /* Check for valid centralized group */
+            char centralized_group[256] = {0};
+            char centralized_group_token[2] = "G:";
+	    
+            if(strncmp(++tmpstr,centralized_group_token,2)==0)
+            {
+                
+                char group_path[PATH_MAX] = {0};
+        
+                sscanf(tmpstr," G:\'%255[^\']\"",centralized_group);
+                
+                if(snprintf(group_path,PATH_MAX,isChroot() ? "/etc/shared/%s" : DEFAULTDIR"/etc/shared/%s",centralized_group) >= PATH_MAX){	
+                    merror("Invalid group name: %.255s... , group path is too large.",centralized_group);
+                    snprintf(response, 2048, "ERROR: Invalid group name: %.255s...\n\n, group path is too large", centralized_group);
+                    SSL_write(ssl, response, strlen(response));
+                    snprintf(response, 2048, "ERROR: Unable to add agent.\n\n");
+                    SSL_write(ssl, response, strlen(response));
+                    SSL_free(ssl);
+                    close(client.socket);
+                    continue;
+                }
+
+                /* Check if group exists */
+                DIR *group_dir = opendir(group_path);
+                if(group_dir){ 
+                    add_to_centralized_group = 1;
+                }
+                else{
+                    merror("Invalid group: %.255s",centralized_group);
+                    snprintf(response, 2048, "ERROR: Invalid group: %s\n\n", centralized_group);
+                    SSL_write(ssl, response, strlen(response));
+                    snprintf(response, 2048, "ERROR: Unable to add agent.\n\n");
+                    SSL_write(ssl, response, strlen(response));
+                    SSL_free(ssl);
+                    close(client.socket);
+                    continue;
+                }
+                closedir(group_dir);
+
+                /*Forward the string pointer G:'........' 2 for G:, 2 for ''*/
+                tmpstr+= 2+strlen(centralized_group)+2;
+            }
+
+            /* Check for IP when client uses -s option */
+            char client_source_ip[IPSIZE + 1] = {0};
+            char client_source_ip_token[3] = "IP:";
+	    
+            if(strncmp(++tmpstr,client_source_ip_token,3)==0)
+            {
+                sscanf(tmpstr," IP:\'%15[^\']\"",client_source_ip);
+                memcpy(srcip,client_source_ip,IPSIZE);
```

Comment:
This instruction overwrites the source IP address but does not overwrite the `use_source_ip` option.